tgui/packages/tgui/components/AnimatedNumber.tsx

```diff
@@ -0,0 +1,187 @@
+ * convergence.
+ */
+const Q = 0.8333;
+
+/**
+ * A small number.
+ */
+const EPSILON = 10e-4;
+
+/**
+ * An animated number label. Shows a number, formatted with an optionally
+ * provided function, and animates it towards its target value.
+ */
+export class AnimatedNumber extends Component<AnimatedNumberProps> {
+  /**
+   * The inner `<span/>` being updated sixty times per second.
+   */
+  ref = createRef<HTMLSpanElement>();
+
+  /**
+   * The interval being used to update the inner span.
+   */
+  interval?: NodeJS.Timeout;
+
+  /**
+   * The current value. This values approaches the target value.
+   */
+  currentValue: number = 0;
+
+  constructor(props: AnimatedNumberProps) {
+    super(props);
+
+    const { initial, value } = props;
+
+    if (initial !== undefined && isSafeNumber(initial)) {
+      this.currentValue = initial;
+    } else if (isSafeNumber(value)) {
+      this.currentValue = value;
+    }
+  }
+
+  componentDidMount() {
+    if (this.currentValue !== this.props.value) {
+      this.startTicking();
+    }
+  }
+
+  componentWillUnmount() {
+    // Stop animating when the component is unmounted.
+    this.stopTicking();
+  }
+
+  shouldComponentUpdate(newProps: AnimatedNumberProps) {
+    if (newProps.value !== this.props.value) {
+      // The target value has been adjusted; start animating if we aren't
+      // already.
+      this.startTicking();
+    }
+
+    // We render the inner `span` directly using a ref to bypass inferno diffing
+    // and reach 60 frames per second--tell inferno not to re-render this tree.
+    return false;
+  }
+
+  /**
+   * Starts animating the inner span. If the inner span is already animating,
+   * this is a no-op.
+   */
+  startTicking() {
+    if (this.interval !== undefined) {
+      // We're already ticking; do nothing.
+      return;
+    }
+
+    this.interval = setInterval(() => this.tick(), SIXTY_HZ);
+  }
+
+  /**
+   * Stops animating the inner span.
+   */
+  stopTicking() {
+    if (this.interval === undefined) {
+      // We're not ticking; do nothing.
+      return;
+    }
+
+    clearInterval(this.interval);
+
+    this.interval = undefined;
+  }
+
+  /**
+   * Steps forward one frame.
+   */
+  tick() {
+    const { currentValue } = this;
+    const { value } = this.props;
+
+    if (isSafeNumber(value)) {
+      // Converge towards the value.
+      this.currentValue = currentValue * Q + value * (1 - Q);
+    } else {
+      // If the value is unsafe, we're never going to converge, so stop ticking.
+      this.stopTicking();
+    }
+
+    if (Math.abs(value - this.currentValue) < EPSILON) {
+      // We're about as close as we're going to get--snap to the value and
+      // stop ticking.
+      this.currentValue = value;
+      this.stopTicking();
+    }
+
+    if (this.ref.current) {
+      // Directly update the inner span, without bothering inferno.
+      this.ref.current.innerText = this.getText();
```

Comment:
`textContent` is a bit more standard way to set the text and might actually have better performance.

Reply:
Taking me back to before I even knew jQuery was a thing. I knew I was getting into a time machine when I signed up for this, but I wasn't expecting to go this far back :)